### PR TITLE
Added pagination on gradebook

### PR DESCRIPTION
**Pull request by amir-qayyum-khan**
Hi

In this PR I have added pagination on gradebook, 20 records per page.
**Need suggestion** on search on gradebook..Do i need to modify it?

This change applied on Student admin page of both instructor and CCX coach.

Issue: https://github.com/mitocw/edx-platform/issues/92
@pdpinch @pwilkins @giocalitri @jamiefolsom 

screen shot 2015-10-02 at 8 06 05 pm

### CODE REVIEW

File: lms/djangoapps/ccx/views.py [hunk 1]
```diff
@@ -7,6 +7,7 @@
 import json
 import logging
 import pytz
+from math import ceil
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: nit: Looking at the organization of the code, this should go after line 12 (in the `from` section)

File: lms/djangoapps/ccx/views.py [hunk 2]
```diff
@@ -22,6 +23,7 @@
 from django.core.validators import validate_email
 from django.http import Http404
 from django.shortcuts import redirect
+from django.core.paginator import Paginator, EmptyPage, PageNotAnInteger
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: nit: this section seems to be sorted, so please move this line in the right position

File: lms/djangoapps/ccx/views.py [hunk 3]
```diff
@@ -554,8 +556,22 @@ def ccx_gradebook(request, course, ccx=None):
             for student in enrolled_students
         ]
 
+        max_students_per_page = 20
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: In my opinion this should go in a global setting or something like that, but I'm not sure how the rest of the code is organized, so maybe just wait for Sarina's input.

File: lms/djangoapps/ccx/views.py [hunk 4]
```diff
@@ -554,8 +556,22 @@ def ccx_gradebook(request, course, ccx=None):
...
+        paginator = Paginator(student_info, max_students_per_page)
+
+        page = request.GET.get('page', 1)
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: no sanity check for weird inputs?

File: lms/djangoapps/ccx/views.py [hunk 5]
```diff
@@ -554,8 +556,22 @@ def ccx_gradebook(request, course, ccx=None):
...
+
+        try:
+            students = paginator.page(page)
+        except PageNotAnInteger:
+            students = paginator.page(1)
+        except EmptyPage:
+            students = paginator.page(paginator.num_pages)
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: Oh, I see you do it here...

File: lms/djangoapps/instructor/views/api.py [hunk 6]
```diff
@@ -2586,8 +2588,22 @@ def spoc_gradebook(request, course_id):
         for student in enrolled_students
     ]
 
+    max_students_per_page = 20
```
Comment (giocalitri) [inline lms/djangoapps/instructor/views/api.py]: well this needs to go definitively to a variable common to the other file

File: lms/djangoapps/instructor/views/api_urls.py [hunk 7]
```diff
@@ -122,6 +122,9 @@
     url(r'^gradebook$',
         'instructor.views.api.spoc_gradebook', name='spoc_gradebook'),
 
+    url(r'^gradebook/(?P<page>[0-9]+)$',
+        'instructor.views.api.spoc_gradebook', name='spoc_gradebook'),
```
Comment (giocalitri) [inline lms/djangoapps/instructor/views/api_urls.py]: why did you add an extra URL for the same view? Isn't it enough to make the extra part optional?

Comment (amir-qayyum-khan) [inline lms/djangoapps/instructor/views/api_urls.py]: I use 2 url declaration because i am getting same issue as http://stackoverflow.com/questions/1842389/reverse-for-with-arguments-and-keyword-arguments-not-found

**Comment by amir-qayyum-khan** [comment]
@giocalitri done

File: lms/djangoapps/instructor/views/api_urls.py (hunk 7 above)
Comment (giocalitri) [inline lms/djangoapps/instructor/views/api_urls.py]: fair enough

**Comment by giocalitri** [comment]
I guess the only thing missing are tests

**Comment by amir-qayyum-khan** [comment]
@giocalitri Do I need to add tests, because i found no coverage issue on 
- paver test_system -t lms/djangoapps/ccx
- paver test_system -t lms/djangoapps/courseware
- paver test_system -t lms/djangoapps/instructor

**Comment by giocalitri** [comment]
Well, you changed the code: I expect to see at least a test that checks the new code. If there is already such a test, just tell me.

**Comment by amir-qayyum-khan** [comment]
@giocalitri there is not now, let me write one, Thanks

**Comment by amir-qayyum-khan** [comment]
Done @giocalitri

File: lms/djangoapps/ccx/views.py [hunk 8]
```diff
@@ -559,8 +563,20 @@ def ccx_gradebook(request, course, ccx=None):
             for student in enrolled_students
         ]
 
+        paginator = Paginator(student_info, settings.MAX_STUDENTS_PER_PAGE)
```
Comment (giocalitri) [inline lms/djangoapps/ccx/views.py]: nit: is `MAX_STUDENTS_PER_PAGE` used only in the gradebook? if so, maybe we should use a specific name, otherwise leave it as is.

**Comment by giocalitri** [comment]
Just one minor comment, otherwise it looks good to me and I think we can open the PR against edx

**Comment by pdpinch** [comment]
moved to edx
